artiq/applets/big_number.py

```diff
@@ -26,54 +31,73 @@ def __init__(self, args, ctl):
         QtWidgets.QStackedWidget.__init__(self)
         self.dataset_name = args.dataset
         self.ctl = ctl
+        self.value = None
+        self.metadata = {}
+
+        font = QtGui.QFont()
+        font.setPointSize(60)
 
-        self.lcd_widget = QResponsiveLCDNumber()
-        self.lcd_widget.setDigitCount(args.digit_count)
-        self.lcd_widget.doubleClicked.connect(self.start_edit)
-        self.addWidget(self.lcd_widget)
+        self.display_widget = QResponsiveLabel()
+        self.display_widget.setFont(font)
+        self.display_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.display_widget.doubleClicked.connect(self.start_edit)
+        self.addWidget(self.display_widget)
 
         self.edit_widget = QCancellableLineEdit()
-        self.edit_widget.setValidator(QtGui.QDoubleValidator())
-        self.edit_widget.setAlignment(QtCore.Qt.AlignRight)
-        self.edit_widget.editCancelled.connect(self.cancel_edit)
+        self.edit_widget.setFont(font)
+        self.edit_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.edit_widget.editCancelled.connect(
+            lambda: self.setCurrentWidget(self.display_widget))
         self.edit_widget.returnPressed.connect(self.confirm_edit)
         self.addWidget(self.edit_widget)
 
-        font = QtGui.QFont()
-        font.setPointSize(60)
-        self.edit_widget.setFont(font)
+        float_regexp = r"(([+-]?\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?)"
+        self.regexp = "^" + float_regexp + r"( [a-zA-Z]*)?$"
 
-        self.setCurrentWidget(self.lcd_widget)
+        self.setCurrentWidget(self.display_widget)
 
     def start_edit(self):
-        # QLCDNumber value property contains the value of zero
-        # if the displayed value is not a number.
-        self.edit_widget.setText(str(self.lcd_widget.value()))
+        if self.value is None:
+            self.edit_widget.setText("---")
+        else:
+            unit = self.metadata.get("unit", "")
+            scale = scale_from_metadata(self.metadata)
+            value = self.value/scale
+            self.edit_widget.setText(f"{value} {unit}".strip())
         self.edit_widget.selectAll()
         self.edit_widget.setFocus()
         self.setCurrentWidget(self.edit_widget)
 
     def confirm_edit(self):
-        value = float(self.edit_widget.text())
-        self.ctl.set_dataset(self.dataset_name, value)
-        self.setCurrentWidget(self.lcd_widget)
-
-    def cancel_edit(self):
-        self.setCurrentWidget(self.lcd_widget)
+        entry = re.match(self.regexp, self.edit_widget.text())
+        if entry is not None:
+            value = float(entry.group(1))
```

Comment:
Datasets can take multiple float / integer types from numpy, this applet will collapse them all to native float. See #2176.

Reply:
Unsure whether this is of concern in this applet, but figured I would comment on it in case.

Reply:
Can you give examples where the type casting to float might fail on this?

Reply:
A `np.float8` type will convert to a `float`. More problematic is an `int` will also be forced to `float`. My recommendation would be to (in the case of an existing dataset) store the type of the value and perform the appropriate type casting before running set_dataset again.